ath79: add support for D-Link DIR-859 A1

hackpascal opened the pull request:
Add support for D-Link DIR-859 A1

Hardware spec of DIR-859 A1:
SoC: QCA9563
DRAM: 64MB DDR2
Flash: 16MB SPI-NOR
Switch: QCA8337N
WiFi 5.8GHz: QCA9880

USB is supported on the PCB but not connected.

Flash instructions:

1. Upgrade the factory.bin through the factory web interface or the u-boot failsafe interface.
   The firmware will boot up correctly for the first time.
   DO NOT POWER OFF the device after OpenWrt has booted. Otherwise the u-boot will enter failsafe mode as the checksum of the firmware has been changed.
2. Upgrade the sysupgrade.bin through the OpenWrt web interface.
   After upgrading completes the u-boot won't complain about the firmware checksum and it's OK to use now.
3. If you powered off the device before upgrading the sysupgrade.bin, just upgrade the factory.bin through the u-boot failsafe interface and then go to step 2.

Support for seama image is also added.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

ynezz commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -0,0 +1,179 @@
+			};
+
+			partition@70000 {
+				label = "firmware";
```
You're missing `compatible` mtd partition splitter property here, it should be probably set to `compatible = "seama";`

ynezz replied:
@hackpascal Thanks. FYI https://github.com/openwrt/openwrt/pull/1586

ynezz commented on target/linux/ath79/image/generic.mk:
```diff
@@ -43,6 +43,28 @@ define Build/elecom-header
 
 endef
 
+define Device/seama
```
@hackpascal Device support belongs to the other commit.

hackpascal replied:
I'll re-organize it.

mkresin commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -0,0 +1,180 @@
+
+&wmac {
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
```
You can add `qca,no-eeprom;` and patch the mac address in the ath9k-eeprom-ahb-18100000.wmac.bin section of base-files/etc/hotplug.d/firmware/10-ath9k-eeprom instead of base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac. Similar to what you are doing for ath10k and most of the boards are doing for ath9k.

Some time ago I asked to do the same for the k2t via mail. Mind to have a look at it?

hackpascal replied:
I'll try for that.

mkresin commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -0,0 +1,180 @@
+		#size-cells = <0>;
+
+		gpio_usb_power {
+			gpio-export,name = "d-link:reset:switch";
```
doesn't really match the gpio_usb_power node name.

hackpascal replied:
That's a mistake. I copied this node from another dts file. This gpio is connected to the RESET_L pin of the switch.

mkresin commented on target/linux/ath79/image/Makefile:
```diff
@@ -66,6 +74,28 @@ define Device/Default
 endef
 
+define Device/seama
+$(Device/Default)
```
`$(Device/Default)`is always included and can be dropped here.

hackpascal replied:
Will delete this line.

mkresin commented on target/linux/ath79/image/Makefile:
```diff
@@ -66,6 +74,28 @@ define Device/Default
+  LOADER_TYPE := bin
+  KERNEL := kernel-bin | append-dtb | relocate-kernel | lzma
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | relocate-kernel | lzma | seama
+  KERNEL_INITRAMFS_SUFFIX = $$(KERNEL_SUFFIX).seama
```
yes, can be done. But I don't see much benefit of doing so as long as we are only able to create a single initramfs image.

hackpascal replied:
This is copied from the ar71xx target. I also think this is useless. I'll remove these lines.

mkresin commented on target/linux/ath79/image/Makefile:
```diff
@@ -66,6 +74,28 @@ define Device/Default
+	check-size $$$$(IMAGE_SIZE)
+  IMAGE/factory.bin := \
+	$$(IMAGE/default) | seama | pad-rootfs | \
+	seama-seal -m "signature=$$$$(SEAMA_SIGNATURE)" | \
```
wouldn't it make more sense to move the `-m "signature=$$$$(SEAMA_SIGNATURE)"` to `Build/seama-seal`

hackpascal replied:
This is also copied from the ar71xx target.

mkresin commented on target/linux/ath79/image/Makefile:
```diff
@@ -66,6 +74,28 @@ define Device/Default
+	$$(IMAGE/default) | seama | pad-rootfs | \
+	seama-seal -m "signature=$$$$(SEAMA_SIGNATURE)" | \
+	check-size $$$$(IMAGE_SIZE)
+  SEAMA_SIGNATURE :=
```
Please add it to the `DEVICE_VARS` instead.

hackpascal replied:
This is also copied from the ar71xx target. I'll change it.

mkresin commented on target/linux/ath79/image/Makefile:
```diff
@@ -66,6 +74,28 @@ define Device/Default
 	append-rootfs | pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
 endef
 
+define Device/seama
```
I would like to see it moved to target/linux/ath79/image/generic.mk for now, as the only image using the device definition is in that file. As soon as other subtargets require the (default) device definition, it should be moved.

hackpascal replied:
OK

mkresin commented on target/linux/ath79/image/Makefile:
```diff
@@ -66,6 +74,28 @@ define Device/Default
+  KERNEL := kernel-bin | append-dtb | relocate-kernel | lzma
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | relocate-kernel | lzma | seama
+  KERNEL_INITRAMFS_SUFFIX = $$(KERNEL_SUFFIX).seama
+  IMAGES := sysupgrade.bin factory.bin
```
`IMAGES += factory.bin` is enough. sysupgrade.bin is already added in Makefile:define Device/Default.

hackpascal replied:
OK

mkresin commented on target/linux/ath79/image/generic.mk:
```diff
@@ -138,6 +138,21 @@ define Device/dlink_dir-825-b1
+  IMAGE_SIZE := 15872k
+  DEVICE_PACKAGES :=  kmod-usb-core kmod-usb2 kmod-ath10k-ct ath10k-firmware-qca988x-ct
+  SEAMA_SIGNATURE := wrgac37_dlink.2013gui_dir859
+  IMAGE/factory.bin := \
```
If you can't use the factory.bin recipe from `Device/seama`, drop the `Device/seama` factory.bin recipe. No need to carry around unused code, which might (or not) be required at some time. It should be added if required.

hackpascal replied:
The factory.bin must be used for the first time upgrading from the factory firmware.
Then upgrade the sysupgrade.bin from OpenWrt to avoid triggering u-boot's firmware checksum validation.

hackpascal replied:
Sorry, misunderstood your words.
Normally we expect the jffs2 end-of-filesystem marker is not covered by the seama header.
Because the seama header has a md5 checksum field. If the jffs2 marker is covered by the header, once OpenWrt booted up the jffs2 marker will be erased and the checksum will become incorrect.
However the d-link's firmware and u-boot require the whole firmware to be covered by the header. The factory.bin will only pass the u-boot's validation for the first time it boots up. The u-boot will refuse to boot the firmware after reboot and enter the failsafe mode. So the sysupgrade.bin must be upgraded in the factory.bin at the first boot.

So, I assume this is a special case for d-link's firmware, and choose to override the recipe instead of replacing the Device/seama.

mkresin commented on target/linux/ath79/image/generic.mk:
```diff
@@ -138,6 +138,21 @@ define Device/dlink_dir-825-b1
 TARGET_DEVICES += dlink_dir-825-b1
 
+define Device/dlink_dir-859-a1
+$(Device/seama)
```
please indent this line as well.

hackpascal replied:
OK

mkresin commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -28,6 +28,10 @@ buffalo,whr-g301n)
 	;;
+dlink,dir-859-a1)
+	ucidef_set_led_switch "internet" "WAN" "d-link:green:internet" "switch0" "0x20"
+	ucidef_set_led_wlan "wlan" "WLAN" "d-link:green:wlan" "phy0tpt"
```
Please set the trigger in the dts file. You should find quite some examples in ath79/dts/.

hackpascal replied:
I think so.

mkresin commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -28,6 +28,10 @@ buffalo,whr-g301n)
 	ucidef_set_led_switch "lan4" "LAN4" "$boardname:green:lan4" "switch0" "0x10"
 	;;
+dlink,dir-859-a1)
+	ucidef_set_led_switch "internet" "WAN" "d-link:green:internet" "switch0" "0x20"
```
As long as we don't have dozen of d-link boards with a similar led configuration (as we have it for tp-link boards), you should follow the `<boardname>:<colour>:<function>` naming scheme for the leds (`dir-859-a1:green:internet` in the dts).

hackpascal replied:
OK

mkresin commented:
I dropped the seama splitter with 6e626477d5 before checking if required by any pending PR. Please add it back.

hackpascal committed:
ath79: re-add seama mtd splitter

The seama splitter will be used by some D-Link devices.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal commented:
@mkresin I've updated the patch. Please review it, thanks.

hackpascal committed:
ath79: image: add seama image support

Add seama image support for generic.mk

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
ath79: add support for D-Link DIR-859 A1

Hardware spec of DIR-859 A1:
SoC: QCA9563
DRAM: 64MB DDR2
Flash: 16MB SPI-NOR
Switch: QCA8337N
WiFi 5.8GHz: QCA9880

USB is supported on the PCB but not connected.

Flash instructions:

1. Upgrade the factory.bin through the factory web interface or the u-boot
   failsafe interface.
   The firmware will boot up correctly for the first time.
   Do not power off the device after OpenWrt has booted. Otherwise the u-boot
   will enter failsafe mode as the checksum of the firmware has been changed.
2. Upgrade the sysupgrade.bin in OpenWrt.
   After upgrading completes the u-boot won't complain about the firmware
   checksum and it's OK to use now.
3. If you powered off the device before upgrading the sysupgrade.bin, just
   upgrade the factory.bin through the u-boot failsafe interface and then goto
   step 2.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal commented:
rebased code to resolve conflicts

mkresin commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -0,0 +1,181 @@
+		wps {
+			label = "dir-859-a1:green:wps";
+			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
https://git.openwrt.org/0d23fd2ab29a66f5d03187db4fac3e396b4f3b62

mkresin commented:
merged, thanks!
